design_docs/oscode/oscode_sundials.md

```diff
@@ -0,0 +1,375 @@
+| Minimum allowed step reduction factor on error test fail     | `OSCStepSetMinReduction()`      | 0.1      |
+| Time step safety factor                                      | `OSCStepSetSafetyFactor()`      | 0.96     |
+| Error fails before MaxEFailGrowth takes effect               | `OSCStepSetSmallNumEFails()`    | 2        |
+| Explicit stability function                                  | `OSCStepSetStabilityFn()`       | none     |
+
+
+## Reference-level explanation
+------------------------------------
+This is the technical portion of the RFC. Explain the design in sufficient detail that:
+
+- Its interaction with other features is clear.
+- It is reasonably clear how the feature would be implemented.
+- Corner cases are dissected by example.
+
+The section should return to the examples given in the previous section, and explain more fully how the detailed proposal makes those examples work.
+
+------------------------------------
+
+The main driver of `OSCODE` will be the function `oscEvolve` with a similar signature to `arkEvolve`
+
+```C
+/*---------------------------------------------------------------
+  oscEvolve:
+
+  This routine is the main driver of OSCODE-based integrators.
+
+  It integrates over a time interval defined by the user, by
+  calling the time step module to do internal time steps.
+
+  The first time that oscEvolve is called for a successfully
+  initialized problem, it computes a tentative initial step size.
+
+  oscEvolve supports two modes as specified by itask: OSC_NORMAL and
+  OSC_ONE_STEP.  In the OSC_NORMAL mode, the solver steps until
+  it reaches or passes tout and then interpolates to obtain
+  y(tout).  In the OSC_ONE_STEP mode, it takes one internal step
+  and returns.  The behavior of both modes can be over-rided
+  through user-specification of osc_tstop (through the
+  *StepSetStopTime function), in which case if a solver step
+  would pass tstop, the step is shortened so that it stops at
+  exactly the specified stop time, and hence interpolation of
+  y(tout) is not required.
+  ---------------------------------------------------------------*/
+int oscEvolve(OSCodeMem osc_mem, realtype tout, N_Vector yout,
+              realtype *tret, int itask);
+```
+
+The `OSCodeMemRec` will be similar to `ARKodeMemRec` containing the information for the estimation using the rk and wkb solver
+
+(TODO:) The example program `example_oscode.cpp` shows an example of the user API for the burst equation.
+
+Users can set:
+
+1. `w` and `g` functions that take in a timestep `t` along with a `void*` argument for their own data
+2. `OSCODEOmegaFn` and `OSCODEGammaFn`, the functors $\omega(t)$ and $\gamma(t)$ which have pointer type
+    - `complextype (OSCODEOmegaFn*)(realtype)`
+3. Options for the rk solver
+    - order
+    - tolerances
+    - butcher tables
+    - Gauss - Lobatto weights
+    - Integral Function
```

Comment:
What's this?

Reply:
Scratch that, I was thinking of `g` but we document that earlier